sql/dd/impl/upgrade/server.cc

```diff
@@ -518,6 +518,48 @@ bool upgrade_help_tables(THD *thd) {
   return false;
 }
 
+/*
+  For SYSTEM table, Run alter tables with target ddse during dd engine change
+*/
+bool fix_system_tables_for_engine_change(THD *thd) {
+  if (dd::execute_query(thd, "USE mysql")) {
+    LogErr(ERROR_LEVEL, ER_DD_UPGRADE_FAILED_FIND_VALID_DATA_DIR);
+    assert(false);
+    return true;
+  }
+
+  // iterate SYSTEM tables and run alter engine=target ddse
+  for (System_tables::Const_iterator it =
+           System_tables::instance()->begin(System_tables::Types::SYSTEM);
+       it != System_tables::instance()->end();
+       it = System_tables::instance()->next(it, System_tables::Types::SYSTEM)) {
+    assert((*it)->property() == System_tables::Types::SYSTEM);
+
+    dd::String_type table_name = (*it)->key().first + "." + (*it)->key().second;
+
+    if (table_name == "mysql.gtid_executed" ||
+        table_name == "mysql.password_history")
+      continue;
+
+    dd::String_type alter_cmd =
+        "alter table " + table_name + " engine=" + get_dd_engine_name();
+    Ed_connection con(thd);
+    LEX_STRING str;
+    lex_string_strmake(thd->mem_root, &str, alter_cmd.c_str(),
+                       alter_cmd.length());
+
+    // ignore ER_NO_SUCH_TABLE due to some tables in System_tables mayn't exists
```

Comment:
As discussed offline, we need to have a list of the tables that are allowed to be missing and allow `ER_NO_SUCH_TABLE` only for them